internal/app/ldap/ldap.go

```diff
@@ -79,54 +96,106 @@ func (l *LdapAuthentication) Authenticate(username string, password string) (int
-	var sb strings.Builder
-	sb.WriteString("Cn=")
-	sb.WriteString(username)
-	sb.WriteString(",")
-	sb.WriteString(cf.BaseDN)
-	err = ldapconn.Bind(sb.String(), password)
-	if err != nil {
-		l.Errorf("can't authentication ")
-		return err
+	ok := false
+	for _, dn := range userdn {
+		err = conn.Bind(dn, password)
+		if err == nil {
+			ok = true
+			break
+		}
+	}
+	if ok {
+		return nil
 	}
-	return nil
+	return errors.New("login fails")
+
 }
-func ldapquery(username string, cf LDAPconf, l glog.Logger) (types.User, error) {
-	ldapconn, err := ldap.Dial("tcp", fmt.Sprintf("%s:%d", cf.LdapAddr, cf.LdapPort))
+func ldapquery(username string, cf LDAPconf, l glog.Logger, conn *ldap.Conn) (types.User, error) {
+	err := conn.Bind(cf.LdapFirstBindUsername, cf.LdapFirstBindPassword)
 	if err != nil {
 		return types.User{}, err
 	}
-	defer ldapconn.Close()
-	err = ldapconn.Bind(cf.LdapFirstBindUsername, cf.LdapFirstBindPassword)
+	searchRequest := ldap.NewSearchRequest(
+		cf.BaseDN,
+		ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
+		fmt.Sprintf("(&(objectClass=*)(cn=%s))", username),
+		[]string{"dn", "mail", "memberOf", "description", "displayName", "cn"},
+		nil,
+	)
+	searchresult, err := conn.Search(searchRequest)
 	if err != nil {
+		l.Errorf("ldap query search fail")
+		return types.User{}, err
+	}
+	if len(searchresult.Entries) == 0 {
+		l.Errorf("no user query")
+		err := errors.New("user does not exist ")
 		return types.User{}, err
 	}
+	for _, entries := range searchresult.Entries {
+		if strlib.IsLower(entries.GetAttributeValue("cn")) {
+			emailaddr := entries.GetAttributeValue("mail")
+			ds := entries.GetAttributeValue("description")
+			// try to get DC name form description value
+			dc := getdcformdescripton(ds, "DC[0-9]+")
+			fulname := entries.GetAttributeValues("displayName")
+			var name string
+			if len(fulname) == 0 {
+				name = ""
+			}
+			name = fulname[0]
+			return types.User{
+				Username:       username,
+				Fullname:       name,
+				DeliveryCenter: dc,
+				Email:          emailaddr,
+			}, nil
+		}
+	}
+	return types.User{}, errors.New("can't get user information form ldap")
+}
+func finduserdn(username string, cf LDAPconf, l glog.Logger, conn *ldap.Conn) ([]string, error) {
+	err := conn.Bind(cf.LdapFirstBindUsername, cf.LdapFirstBindPassword)
+	if err != nil {
+		l.Errorf("can't first bind to ldap database to query user dn ")
+		return nil, err
+	}
+	l.Infof("query user by basedn : %s", cf.BaseDN)
 	searchRequest := ldap.NewSearchRequest(
 		cf.BaseDN,
 		ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
 		fmt.Sprintf("(&(objectClass=*)(cn=%s))", username),
-		[]string{"dn", "mail", "memberOf", "description", "displayName"},
+		[]string{"dn"},
 		nil,
 	)
-	searchrequest, err := ldapconn.Search(searchRequest)
+	searresult, err := conn.Search(searchRequest)
 	if err != nil {
-		l.Errorf("search fail")
-		return types.User{}, err
+		l.Errorf("search dn for login fail")
```

Comment:
with just a message like "search dn for login fail", you would have no idea what happen and have no chance to troubleshoot from the log. For all error log, please make sure you log the error accordingly.